### PR TITLE
Parse optional type info in JVC files

### LINKED ISSUE #1509: Parse optional type information in JVC files
It would be useful for testing transformations that manipulate type information to have the possibility of optionally specifying (partial) types.

Examples
----------
* `type list;`
  Declare a `list` type symbol (to be used later with TypeConstr).
* `def f : Nat -> Nat := body;`
* `constr cons : * -> * -> *;`
* `constr cons : (A : Type) -> A -> list A -> list A;`
  These should add appropriate type information to the InfoTable. It's useful to be able to specify partial type information with the dynamic type (`*`).
* `\x : Nat . f x x`
  The `Nat` should land in the TypeInfo of the BinderInfo of the Lambda.

### PR DESCRIPTION
**Pull request by lukaszcz**
Parsing of optional type info for JuvixCore source files. This will help with debugging transformations that manipulate type information.

With this change, constructors can no longer have an unspecified type -- they must belong to exactly one inductive type.

Fixes #1509

### CODE REVIEW

**Review by janmasrovira — DISMISSED** [review]

File: src/Juvix/Compiler/Core/Translation/FromSource.hs [hunk 1]
```diff
@@ -538,17 +651,42 @@ parseLocalName = parseWildcardName <|> parseIdentName
...
+  where
+    lambdaName =
+      parens
+        ( do
+            n <- parseLocalName
+            kw kwColon
+            ty <- expr varsNum vars
+            return (n, Just ty)
+        )
+        <|> (parseLocalName >>= \n -> return (n, Nothing))
```
Comment (janmasrovira) [inline src/Juvix/Compiler/Core/Translation/FromSource.hs:689]: `(\n -> (n, Nothing)) <$> parseLocalName`

File: src/Juvix/Compiler/Core/Extra/Base.hs [hunk 2]
```diff
@@ -119,11 +143,33 @@ mkDynamic' = mkDynamic Info.empty
...
+
+typeTarget :: Type -> Type
+typeTarget = fst . unfoldType
+
+isDynamic :: Type -> Bool
+isDynamic = \case
+  NDyn {} -> True
+  _ -> False
+
+expandType :: [Type] -> Type -> Type
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Extra/Base.hs:162]: Done

File: src/Juvix/Compiler/Core/Translation/FromSource.hs [hunk 3]
```diff
@@ -92,22 +75,75 @@ guardSymbolNotDefined sym err = do
...
+  Symbol ->
+  BuiltinDataTag ->
+  Text ->
+  Type ->
+  Interval ->
+  Sem r ConstructorInfo
+createBuiltinConstr sym btag nameTxt ty i = do
+  name <- freshName KNameConstructor nameTxt i
+  let n = builtinConstrArgsNum btag
+   in return $
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Translation/FromSource.hs:89]: Done

File: src/Juvix/Compiler/Core/Translation/FromSource.hs [hunk 4]
```diff
@@ -128,37 +164,51 @@ statementDef = do
...
         (parseFailure off ("duplicate definition of: " ++ fromText txt))
-      parseDefinition sym
-    Just (IdentTag {}) ->
+      tab <- lift getInfoTable
+      let fi = fromMaybe impossible $ HashMap.lookup sym (tab ^. infoIdentifiers)
+          ty = fi ^. identifierType
+      parseDefinition sym ty
+    Just (IdentInd {}) ->
+      parseFailure off ("duplicate identifier: " ++ fromText txt)
+    Just (IdentConstr {}) ->
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Translation/FromSource.hs:177]: Done

File: src/Juvix/Compiler/Core/Translation/FromSource.hs [hunk 5]
```diff
@@ -212,10 +290,47 @@ expr ::
   Members '[Reader ParserParams, InfoTableBuilder, NameIdGen] r =>
   -- current de Bruijn index, i.e., the number of binders upwards
   Index ->
-  -- reverse de Bruijn indices
+  -- reverse de Bruijn indices (de Bruijn levels)
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Translation/FromSource.hs:293]: I created a type synonym. Creating a newtype is inconvenient for the same reason that creating a newtype for Index is: you want to do arithmetic on and then the conversions become cumbersome.

File: src/Juvix/Compiler/Core/Translation/FromSource.hs [hunk 6]
```diff
@@ -212,10 +290,47 @@ expr ::
   Members '[Reader ParserParams, InfoTableBuilder, NameIdGen] r =>
   -- current de Bruijn index, i.e., the number of binders upwards
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Translation/FromSource.hs:291]: Done

File: src/Juvix/Compiler/Core/Translation/FromSource.hs (hunk 1 above)
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Translation/FromSource.hs:689]: Okay, I'll change it. Though I never quite got the point of trying so hard to be so very terse and "point-free" and use all those funny operator symbols. It seems that's what HLint and experienced Haskell programmers recommend. But it makes Haskell look like Perl and hard to read sometimes.

Comment (janmasrovira) [reply src/Juvix/Compiler/Core/Translation/FromSource.hs]: I would say `<$>` is common in the haskell ecosystem

**Review by janmasrovira — APPROVED** [review]

**Pull request merged**